1.regression_modelling/scripts/3.find_sc_crops.py

```diff
@@ -0,0 +1,270 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Generate single-cell crops for min/max montages of the top feature (Zernike phase 2,2) for both stains
+
+# ## Import libraries
+
+# In[1]:
+
+
+import pathlib
+from pprint import pprint
+from typing import List
+
+import cv2
+import pandas as pd
+
+
+# ## Set functions for processing and generating single-cell crops
+
+# In[2]:
+
+
+# Function for formatting min/max row data frames into dictionaries
+def create_sc_dict(dfs: List[pd.DataFrame], names: List[str]) -> dict:
+    """Format lists of data frames and names into a dictionary with all relevant metadata to find single-cell images.
+
+    Args:
+        dfs (List[pd.DataFrame]): List of data frames each containing a single cell and relevant metadata.
+        names (List[str]): List of names corresponding to the data frames.
+
+    Returns:
+        dict: Dictionary containing info relevant for finding single-cell crops.
+    """
+    sc_dict = {}
+    for df, name in zip(dfs, names):
+        for i, (_, row) in enumerate(df.iterrows()):
+            if i >= 5:  # Limit to 5 rows
+                break
+            key = f"{name}_{i+1}"  # Create key with incrementing number
+            sc_dict[key] = {
+                "plate": row["Metadata_Plate"],
+                "well": row["Metadata_Well"],
+                "site": row["Metadata_Site"],
+                "location_center_x": row["Metadata_Nuclei_Location_Center_X"],
+                "location_center_y": row["Metadata_Nuclei_Location_Center_Y"],
+            }
+    return sc_dict
+
+
+# In[3]:
+
+
+# Function for generating and saving single-cell crops per channel as PNGs
+def generate_sc_crops(
+    sc_dict: dict,
+    images_dir: pathlib.Path,
+    output_img_dir: pathlib.Path,
+    crop_size: int,
+) -> None:
+    """Using a dictionary with single-cell metadata info per image set, single-cell crops per channel are generated 
+    and saved as PNGs in an image set folder.
+
+    Args:
+        sc_dict (dict): Dictionary containing info relevant for finding single-cell crops.
+        images_dir (pathlib.Path): Directory where illumination corrected images are found.
+        output_img_dir (pathlib.Path): Main directory to save each image set single-cell crops
+        crop_size (int): Size of the box in pixels (example: setting crop_size as 250 will make a 250x250 pixel crop around the single-cell center coordinates)
+    """
+    for key, info in sc_dict.items():
+        # Initialize a list to store file paths for every image set
+        file_paths = []
+
+        # Create file paths with well, site, and channel
+        for i in range(3):  # Update the range to start from 0 and end at 2
+            filename = f"{images_dir}/{info['plate']}/{info['plate']}_{info['well']}_{info['site']}_CH{i}_Z09_illumcorrect.tiff"
```

Comment:
This is organized but stressful haha. Good job managing chaos into organization.

Reply:
Haha thank you! 😄